x-pack/qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/TokenBackwardsCompatibilityIT.java

```diff
@@ -58,34 +59,40 @@ public void testGeneratingTokenInOldCluster() throws Exception {
                 "{\n" +
                 "    \"token\": \"" + token + "\"\n" +
                 "}");
-        client().performRequest(indexRequest2);
+        Response indexResponse2 = client().performRequest(indexRequest2);
+        assertOK(indexResponse2);
     }
 
-    public void testTokenWorksInMixedOrUpgradedCluster() throws Exception {
-        assumeTrue("this test should only run against the mixed or upgraded cluster",
-                CLUSTER_TYPE == ClusterType.MIXED || CLUSTER_TYPE == ClusterType.UPGRADED);
+    public void testTokenWorksInMixedCluster() throws Exception {
+        // Verify that an old token continues to work during all stages of the rolling upgrade
+        assumeTrue("this test should only run against the mixed cluster", CLUSTER_TYPE == ClusterType.MIXED);
         Request getRequest = new Request("GET", "token_backwards_compatibility_it/_doc/old_cluster_token1");
         Response getResponse = client().performRequest(getRequest);
         assertOK(getResponse);
         Map<String, Object> source = (Map<String, Object>) entityAsMap(getResponse).get("_source");
         assertTokenWorks((String) source.get("token"));
     }
 
-    public void testMixedCluster() throws Exception {
+    public void testInvalidatingTokenInMixedCluster() throws Exception {
+        // Verify that we can invalidate a token in a mixed cluster
         assumeTrue("this test should only run against the mixed cluster", CLUSTER_TYPE == ClusterType.MIXED);
-        assumeTrue("the master must be on the latest version before we can write", isMasterOnLatestVersion());
         Request getRequest = new Request("GET", "token_backwards_compatibility_it/_doc/old_cluster_token2");
-
         Response getResponse = client().performRequest(getRequest);
+        assertOK(getResponse);
         Map<String, Object> source = (Map<String, Object>) entityAsMap(getResponse).get("_source");
-        final String token = (String) source.get("token");
-        assertTokenWorks(token);
-
+        String token = (String) source.get("token");
+        // The token might be already invalidated by running testInvalidatingTokenInMixedCluster in a previous stage
```

Comment:
So we might run mixed cluster tests more than one time?

Reply:
`testInvalidatingTokenInMixedCluster` runs on mixed clusters, so it runs on `oneThirdUpgraded `and `twoThirdUpgraded`.